README.md
AngularLogExtender
==================
[![Build Status](https://travis-ci.org/lwhiteley/AngularLogExtender.svg?branch=master)](https://travis-ci.org/lwhiteley/AngularLogExtender)
[![Coverage Status](https://coveralls.io/repos/github/ferronrsmith/AngularLogExtender/badge.svg?branch=master)](https://coveralls.io/github/ferronrsmith/AngularLogExtender?branch=master)

This is an extension of the Angular $log functionality.
It uses the native $decorator to push the $log pass its capabilities and provide new
functionality such as configuring the $log for different environments such as
production and development.

## How to Use

Please view our [wiki](https://github.com/lwhiteley/AngularLogExtender/wiki) for detailed documentation.


### Notes

You can include the module in your AngularJS Application and it does all the work immediately.
Methods native to the log extender are not publicly available in your AngularJs
Application so this extension can be used as a standalone plugin.
Advanced configurations can be done to make the $log service fit
your personal development style. Log methods are now colour coded by default.


### Install with bower

Now offers bower support.

`bower install --save angular-logex `

Add the script to your index.html:
```javascript
<script src="/bower_components/angular-logex/dist/log-ex-unobtrusive.js"></script>
```

## Credits
To view the blog this module was extended from and inspired by, go to
http://solutionoptimist.com/2013/10/07/enhance-angularjs-logging-using-decorators/

## Development

If you wish to contribute, Please read the `Develop.md` file.

Feel Free to make your own contributions to this module so we can make it better :)

====================
License: MIT |   Made with :green_heart:


CHANGELOG.md

0.0.13 / 2016-09-21
===================

  * added new color support (#49)

0.0.12 / 2016-09-20
===================

  * fixed :bug: whereby the original logged object was being modified by the
    filter function

0.0.11-3+sha.c6b11f5 / 2016-03-07
=================================

  * NPM Package updates
  * Added Angular 1.5.x to test suite
  * Reformat all code

0.0.10-3+sha.c8ba77d / 2015-04-12
=================================

  * Release 0.0.10-3+sha.c8ba77d
  * bump tagrelease version
  * Updates to package.json and libs
  * added new angular libraries to the testing cycle to ensure nothin has broken. the latest 1.2.x, 1.3.x & 1.4.x beta libraries has been added. closes #40
  * Merge pull request #31 from lwhiteley/feature/developer
  * added more details to develop.md for developers & also updated wiki. closes #31
  * Merge pull request #28 from lwhiteley/jshint
  * Update dist task to lint dist file after generation. resolves #28
  * Merge pull request #30 from lwhiteley/readme
  * remove detailed docs from readme. closes #30
  * Merge pull request #29 from lwhiteley/feature/bump
  * Used proper semantic versioning library for numbering. closes #29

0.0.10-2+sha.c8ba77d / 2014-09-22
=================================

  * Release 0.0.10-2+sha.c8ba77d
  * Bumping version to 0.0.10-2.
  * Bumping version to 0.0.10-2.
  * Bumping version to 0.0.10-2.
  * Bumping version to 0.0.10-1.
  * Bumping version to 0.0.10-0.

v0.0.9-sha.6871eea / 2014-09-13
===============================

  * fixed jshint, lint issues & upgraded packages. closes #26

v0.0.9 / 2014-09-19
===================

  * Release v0.0.9
  * Merge pull request #25 from lwhiteley/develop
  * Merge pull request #22 from lwhiteley/logfilter
  * Added support filter sensitive data masking. closes #22
  * Merge pull request #24 from waffle-iron/master
  * Update README.md
  * Update .travis.yml
  * add waffle.io badge

0.0.8-sha.009021d / 2014-09-19
==============================

  * Merge pull request #25 from lwhiteley/develop
  * Merge pull request #22 from lwhiteley/logfilter
  * Added support filter sensitive data masking. closes #22
  * Merge pull request #24 from waffle-iron/master
  * Update README.md
  * Update .travis.yml
  * add waffle.io badge

0.0.6-sha.101ac4a / 2014-02-01 
==================

 * cleanup before release 0.0.7
 * Update known keys return description
 * Update readme and develop documentation
 * Add knownKeys validation for colorCss
 * edit test description
 * Add length validation to colorCss validation
 * edit overrideLogPrefixTest to return customLogPrefix to original state after test
 * add memorization for default logprefix rules
 * add clean task for coverage folder
 * update grunt file with proper indent
 * update clean config to support for multi clean config
 * add minify process for module
 * Add param descriptions
 * Add param descriptions for declarations
 * change for-in to foreach in overrideLogMethodColors method
 * fix spelling and wording of descriptions
 * move provider method descriptions to above method definitions
 * Add overrideLogMEthodColors method and tests
 * Move setLogMethodColor tests to a different file
 * Update readme wording
 * add setLogMethodColor and tests
 * generate release
 * CleanUp - fixed linting issues with methods - remove commented out code - remove unsued variables - fixed formatting of methods to match jslint standards
 * - generate new release - fix grammer issues - added test case for supplant catch block - clean up formatting and parameter passing - correct param naming - set optional types for certain method
 * update documentation
 * update bower settings
 * Update documentation with new config
 * Update documentation
 * Update default colors
 * Update createlogobj method and spec
 * Fix incorrect spelling and add util function
 * Update documentation
 * Add default colors for each log method
 * Update documentation
 * Update typeof references to use itypeof
 * Add tests for supplant
 * Update function names
 * Add string check for supplant pattern
 * update prepareLogFn method and spec
 * Add function description for canColorizee
 * Update documentation
 * Fix spelling error
 * Add function description for canTemplate
 * Implement string template functionality
 * Add documentation
 * Add documentation for colorify
 * Update prepareLogFn tests
 * Enhance colorify method
 * Add supplant method and spec
 * Update colorify tests
 * Update colorify tests
 * Update colorify to validate css string
 * Add itypeof method and tests
 * add colorify method
 * Update dist file
 * Add methods for Colorify supported browser detection
 * added  version testing
 * test cleanup
 * update changelog
 * added jsonlint

0.0.6-sha.c52a8fd / 2014-01-11 
==================

 * added jsonlint

0.0.5-sha.329140e / 2014-01-11 
==================

 * Update Documentation to remove future implementation that was achieved
 * connect bower to dist generation
 * Update Documentation to showcase upcoming bower support
 * Update bower ignore list
 * Add bower support
 * release 0.0.5

0.0.4-sha.a415250 / 2014-01-10 
==================

 * remove unecessary declaration
 * update submodule
 * snapshot release
 * release 0.0.4

0.0.3-sha.665adf9 / 2014-01-10 
==================

 * Update Documentation
 * Update Documentation
 * Update Documentation
 * Update Documentation
 * Update Documentation
 * Remove constant reference
 * Shorten module name to log.ex.uo
 * Replace plain text with constant references
 * Update documentation to give a more accurate description of the module
 * Update Documentation
 * Update Documentation to showcase added configurations
 * Update Documentation to showcase constants
 * Update Documentation and add logMethod constant for external use
 * updated config for lcov
 * more context
 * new dist
 * remove readme dup
 * fixed readme issue
 * updated documentation
 * update submodule
 * update submodule

Gruntfile.js
var util = require('./config/grunt/utils.js');
var files = require('./config/files');

module.exports = function (grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  var port = process.env.PORT || 3000;

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    appVersion: util.getVersion().version,
    meta: {
      files: ['Gruntfile.js', 'dist/*.js', 'test/**/*.js', 'src/**/*.js', 'config/**/*.js'],
      dist: ["dist/*.js"],
      port: port
    },
    jshint: {
      files: ['<%= meta.files %>', 'package.json', 'bower.json'],
      dist: ['<%= meta.dist %>'],
      options: {
        // options here to override JSHint defaults
        jshintrc: '.jshintrc'
      }
    },
    jsbeautifier: {
      files: '<%= meta.dist %>'
    },
    concat: {
      options: {
        separator: '\n',
        process: function (src) {
          var APP_VERSION = util.getVersion();
          return src
            .replace(/%VERSION%/g, APP_VERSION.version)
            .replace(/%WEBSITE%/g, APP_VERSION.website)
            .replace(/%LICENSE%/g, APP_VERSION.license)
            .replace(/%CONTRIBUTOR%/g, APP_VERSION.contributor)
            .replace(/%APP_NAME%/g, APP_VERSION.appname)
            .replace(/%DESCRIPTION%/g, APP_VERSION.description)
            .replace(/%YEAR%/g, new Date().getFullYear())
            .replace(/%SUPPORTED_VERSIONS%/g, APP_VERSION.supported_versions.join(", "));
        }
      },
      minify: {
        src: [
          'src/header.js',
          'dist/log-ex-unobtrusive.min.js'
        ],
        dest: 'dist/log-ex-unobtrusive.min.js'
      },
      dist: {
        src: [
          'src/header.js',
          'src/module.prefix',
          'src/declarations.js',
          // <--------- additional functionality start
          'src/addins/defaults.js',
          'src/addins/helpers.js',
          'src/addins/color.js',
          'src/addins/supplant.js',
          'src/addins/masking.js',
          // < ----------------
          'src/enhanceObj/obj.prefix',
          'src/enhanceObj/main.js',
          // <------- eLogger Start
          'src/eLogger/obj.prefix',
          'src/eLogger/logger.js',
          'src/eLogger/globals.js',
          'src/eLogger/obj.suffix',
          // <------- eLogger End
          // <------- extras Start
          'src/extras/extras.js',
          // <------- extras End
          'src/enhanceObj/globals.js',
          'src/enhanceObj/obj.suffix',
          // < ----------------
          'src/module.suffix',
          // <--------- provider func start
          'src/providerFunc/main.js',
          'src/providerFunc/provider.suffix'
        ],
        dest: 'dist/log-ex-unobtrusive.js'
      }
    },
    karma: {
      options: {
        singleRun: true,
        browsers: ['PhantomJS']
      },
      '1.0.x': {
        options: {
          files: files.getAngularFiles('1.0').concat(files.libs, files.tests('1.0'))
        },
        configFile: 'config/karma.conf.js'
      },
      '1.1.x': {
        options: {
          files: files.getAngularFiles('1.1').concat(files.libs, files.tests('1.1'))
        },
        configFile: 'config/karma.conf.js'
      },
      '1.1.2': {
        options: {
          files: files.getAngularFiles('1.1.2').concat(files.libs, files.tests('1.1.2'))
        },
        configFile: 'config/karma.conf.js'
      },
      '1.2.x': {
        options: {
          files: files.getAngularFiles('1.2').concat(files.libs, files.tests('1.2'))
        },
        configFile: 'config/karma.conf.js'
      },
      '1.3.x': {
        options: {
          files: files.getAngularFiles('1.3').concat(files.libs, files.tests('1.3'))
        },
        configFile: 'config/karma.conf.js'
      },
      '1.4.x': {
        options: {
          files: files.getAngularFiles('1.4').concat(files.libs, files.tests('1.4'))
        },
        configFile: 'config/karma.conf.js'
      },
      latest: {
        options: {
          files: files.getAngularFiles().concat(files.libs, files.tests('latest'))
        },
        configFile: 'config/karma.conf.js'
      },
      performance: {
        options: {
          files: files.getAngularFiles().concat(files.libs, ['test/perf/*.js'])
        },
        configFile: 'config/karma.conf.js'
      },
      coverage: {
        configFile: 'config/karma.lcov.conf.js'
      }
    },
    watch: {
      files: '<%= meta.files %>',
      tasks: ['test'],
      options: {
        spawn: false
      }
    },
    coveralls: {
      options: {
        coverageDir: 'lcov'
      }
    },
    bump: {
      options: {
        filepaths: ['package.json', 'bower.json'],
        syncVersions: true,
        commit: false,
        commitMessage: 'Bumping version to {%= version %}.',
        tag: false,
        tagName: 'v{%= version %}',
        tagMessage: 'Version {%= version %}',
        tagPrerelease: false
      }
    },
    shell: {
      changelog: {
        command: 'git changelog --tag <%= appVersion %>'
      }
    },
    clean: {
      dist: ['dist/'],
      cover: ['coverage/']
    },
    minified: {
      files: {
        src: [
          'dist/log-ex-unobtrusive.js'
        ],
        dest: 'dist/'
      },
      options: {
        sourcemap: true,
        allinone: false,
        ext: '.min.js'
      }
    },
    express: {
      options: {
        port: port
      },
      dev: {
        options: {
          script: 'sample_app/server.js',
          debug: false
        }
      }
    },
    open: {
      server: {
        path: 'http://localhost:<%= meta.port %>'
      }
    },
    copy: {
      files: {
        src: 'dist/*',
        dest: 'sample_app/app/'
      }
    },
    tagrelease: {
      file: 'package.json',
      commit: true,
      prefix: '',
      annotate: false
    }
  });

  grunt.registerTask('test', [
    'clean:cover',
    'jshint:files',
    'karma'
  ]);
  grunt.registerTask('minify', ['minified', 'concat:minify']);
  grunt.registerTask('dist', ['clean:dist', 'test', 'concat:dist', 'jsbeautifier', 'minify', 'jshint:dist']);
  grunt.registerTask('dev-release', ['bump:pr:git', 'dist', 'tagrelease']);
  grunt.registerTask('release', ['bump:patch', 'dist', 'tagrelease']);
  grunt.registerTask('changelog', ['shell:changelog']);
  grunt.registerTask('serve', ['copy:files', 'express:dev', 'open', 'watch']);
  grunt.registerTask('default', ['test', 'coveralls']);
};
